daemon/containerd/image_delete.go

```diff
@@ -53,123 +53,156 @@ import (
 //
 // TODO(thaJeztah): image delete should send prometheus counters; see https://github.com/moby/moby/issues/45268
 func (i *ImageService) ImageDelete(ctx context.Context, imageRef string, force, prune bool) ([]imagetypes.DeleteResponse, error) {
-	parsedRef, err := reference.ParseNormalizedNamed(imageRef)
-	if err != nil {
-		return nil, err
+	var c conflictType
+	if !force {
+		c |= conflictSoft
 	}
 
-	img, err := i.resolveImage(ctx, imageRef)
+	img, all, err := i.resolveAllReferences(ctx, imageRef)
 	if err != nil {
 		return nil, err
 	}
 
-	imgID := image.ID(img.Target.Digest)
-
-	explicitDanglingRef := strings.HasPrefix(imageRef, imageNameDanglingPrefix) && isDanglingImage(img)
-	if isImageIDPrefix(imgID.String(), imageRef) || explicitDanglingRef {
-		return i.deleteAll(ctx, img, force, prune)
-	}
-
-	singleRef, err := i.isSingleReference(ctx, img)
-	if err != nil {
-		return nil, err
-	}
-	if !singleRef {
-		err := i.client.ImageService().Delete(ctx, img.Name)
+	var imgID image.ID
+	if img == nil {
+		if len(all) == 0 {
+			return []imagetypes.DeleteResponse{}, nil
```

Comment:
This should be unreachable, as `resolveAllReferences` would return `ErrImageDoesNotExist`? Shouldn't this also return `ErrImageDoesNotExist` to match that behavior?

https://github.com/moby/moby/blob/04526ea0fe85acdf99cf6d77bf065c72892a641e/daemon/containerd/image_delete_test.go#L30-L33

Reply:
It definitely should be unreachable but if it is an error is easier to follow than a panic or no-op.